std/machines/hash/keccakf16_memory.asm

```diff
@@ -0,0 +1,876 @@
+    link if final_step ~> mem.mstore(addr_h[26], addr_l[26], time_step, a_prime_prime[53], a_prime_prime[52]);
+    link if final_step ~> mem.mstore(addr_h[27], addr_l[27], time_step, a_prime_prime[59], a_prime_prime[58]);
+    link if final_step ~> mem.mstore(addr_h[28], addr_l[28], time_step, a_prime_prime[57], a_prime_prime[56]);
+    link if final_step ~> mem.mstore(addr_h[29], addr_l[29], time_step, a_prime_prime[63], a_prime_prime[62]);
+    link if final_step ~> mem.mstore(addr_h[30], addr_l[30], time_step, a_prime_prime[61], a_prime_prime[60]);
+    link if final_step ~> mem.mstore(addr_h[31], addr_l[31], time_step, a_prime_prime[67], a_prime_prime[66]);
+    link if final_step ~> mem.mstore(addr_h[32], addr_l[32], time_step, a_prime_prime[65], a_prime_prime[64]);
+    link if final_step ~> mem.mstore(addr_h[33], addr_l[33], time_step, a_prime_prime[71], a_prime_prime[70]);
+    link if final_step ~> mem.mstore(addr_h[34], addr_l[34], time_step, a_prime_prime[69], a_prime_prime[68]);
+    link if final_step ~> mem.mstore(addr_h[35], addr_l[35], time_step, a_prime_prime[75], a_prime_prime[74]);
+    link if final_step ~> mem.mstore(addr_h[36], addr_l[36], time_step, a_prime_prime[73], a_prime_prime[72]);
+    link if final_step ~> mem.mstore(addr_h[37], addr_l[37], time_step, a_prime_prime[79], a_prime_prime[78]);
+    link if final_step ~> mem.mstore(addr_h[38], addr_l[38], time_step, a_prime_prime[77], a_prime_prime[76]);
+    link if final_step ~> mem.mstore(addr_h[39], addr_l[39], time_step, a_prime_prime[83], a_prime_prime[82]);
+    link if final_step ~> mem.mstore(addr_h[40], addr_l[40], time_step, a_prime_prime[81], a_prime_prime[80]);
+    link if final_step ~> mem.mstore(addr_h[41], addr_l[41], time_step, a_prime_prime[87], a_prime_prime[86]);
+    link if final_step ~> mem.mstore(addr_h[42], addr_l[42], time_step, a_prime_prime[85], a_prime_prime[84]);
+    link if final_step ~> mem.mstore(addr_h[43], addr_l[43], time_step, a_prime_prime[91], a_prime_prime[90]);
+    link if final_step ~> mem.mstore(addr_h[44], addr_l[44], time_step, a_prime_prime[89], a_prime_prime[88]);
+    link if final_step ~> mem.mstore(addr_h[45], addr_l[45], time_step, a_prime_prime[95], a_prime_prime[94]);
+    link if final_step ~> mem.mstore(addr_h[46], addr_l[46], time_step, a_prime_prime[93], a_prime_prime[92]);
+    link if final_step ~> mem.mstore(addr_h[47], addr_l[47], time_step, a_prime_prime[99], a_prime_prime[98]);
+    link if final_step ~> mem.mstore(addr_h[48], addr_l[48], time_step, a_prime_prime[97], a_prime_prime[96]);
+
+    // ------------- End memory read / write ---------------
+
+    // Adapted from Plonky3 implementation of Keccak: https://github.com/Plonky3/Plonky3/tree/main/keccak-air/src
+
+    std::check::assert(std::field::modulus() >= 65535, || "The field modulo should be at least 2^16 - 1 to work in the keccakf16 machine.");
```

Comment:
can use `require_bits` like the non-memory version

Reply:
Done.